packages/compiler/src/experimental/mutators.ts

```diff
@@ -0,0 +1,389 @@
+import { Program } from "../core/program.js";
+import { isArrayModelType } from "../core/type-utils.js";
+import {
+  Decorator,
+  DecoratorFunction,
+  Enum,
+  EnumMember,
+  FunctionParameter,
+  FunctionType,
+  Interface,
+  IntrinsicType,
+  Model,
+  ModelProperty,
+  Namespace,
+  ObjectType,
+  Operation,
+  Projection,
+  Scalar,
+  ScalarConstructor,
+  StringTemplate,
+  StringTemplateSpan,
+  SyntaxKind,
+  TemplateParameter,
+  Tuple,
+  Type,
+  Union,
+  UnionVariant,
+} from "../core/types.js";
+import { CustomKeyMap } from "../emitter-framework/custom-key-map.js";
+import { $doc, isVisible } from "../lib/decorators.js";
+import { Realm } from "./realm.js";
+import { createTypeFactory } from "./type-factory.js";
+
+export type MutatorRecord<T extends Type> =
+  | {
+      filter?: MutatorFilterFn<T>;
+      mutate: MutatorFn<T>;
+    }
+  | {
+      filter?: MutatorFilterFn<T>;
+      replace: MutatorReplaceFn<T>;
+    }
+  | MutatorFn<T>;
+
+export interface MutatorFn<T extends Type> {
+  (sourceType: T, clone: T, program: Program, realm: Realm): void;
+}
+
+export interface MutatorFilterFn<T extends Type> {
+  (sourceType: T, program: Program, realm: Realm): boolean | MutatorFlow;
+}
+
+export interface MutatorReplaceFn<T extends Type> {
+  (sourceType: T, clone: T, program: Program, realm: Realm): Type;
+}
+
+export interface Mutator {
+  name: string;
```

Comment:
how is the name really used, is it just mostly for debugging?

Reply:
Yes, it is for debugging only